File: 21June2011-FunctionalDojo/multiples.py
// Thomas Parslow: tom@almostobsolete.net @almostobsolete

from itertools import *

def multiples32upTo(n):
    return sum([x for x in xrange(1,n) if x % 3 == 0 or x % 5 == 0])

def multiples32upToB(n):
    return sum(x for x in xrange(1,n) if x % 3 == 0 or x % 5 == 0)

def multiples32upToInf(n):
    return sum(takewhile(lambda x: x < n, (x for x in count(1) if x % 3 == 0 or x % 5 == 0)))

def imperativeVersion(n):
    s = 0
    for n in xrange(n):
        if n % 3 == 0 or n % 5 == 0:
            s = s + n
    return s

print multiples32upTo(1000)
print multiples32upToInf(1000)
print imperativeVersion(1000)

File: 21June2011-FunctionalDojo/multiples.js
// Thomas Parslow: tom@almostobsolete.net @almostobsolete

function sum(list) {
    return _.reduce(list, function (x,y) {return x+y}, 0)
}

function multiples (n) {
    return sum(_.select(_.range(1, 1000), function (x) {return x % 3 == 0 || x % 5 == 0}));
}

alert(multiples(1000));



function multiples2 (from, until) {
    if (from > until) return 0;
    return multiples2(from+1, until) + (from % 3 || from % 5) ? from : 0
}



alert(multiples2(1000)); // Will break the recursion limit